README.md

```diff
@@ -182,3 +182,34 @@ There are some cases defined in the `src/test/index.test.ts` file. After cloning
 pnpm install
 pnpm test
 ```
+
+## Performance Test
+
+We simulate creating a Monorepo and generate a dependency DAG for it. Assume that each project is a node and the dependency relation between any two projects is a directed edge. We also provide a generateDAG function that supports customizing the number of generated nodes and edges for a DAG. You can modify the `testConfig` in `src/performance/index.ts` to customize your own test complexity, then execute the following command to start performance testing
+
+```
+pnpm install
+pnpm test:performance
+```
+
+### Performance report：
+
+-   Testing environment:
+    -   8 \* [Intel(R) Xeon(R) Gold 5118 CPU @ 2.30GHz]
+    -   MemTotal: 40GB
+    -   Node, Typescript
+
+| nodeCount | edgeCount | pathCountA | pathCountB | executeTime |
+| --------- | --------- | ---------- | ---------- | ----------- |
+| 1000      | 5000      | 1          | 1          | 0.489s      |
+| 1000      | 5000      | 10         | 10         | 0.425s      |
+| 1000      | 5000      | 100        | 100        | 0.448s      |
+| 1000      | 5000      | 1000       | 1000       | 0.845s      |
+| 2000      | 10000     | 1          | 1          | 0.913s      |
+| 2000      | 10000     | 10         | 10         | 0.936s      |
+| 2000      | 10000     | 100        | 100        | 1.072s      |
+| 2000      | 10000     | 1000       | 1000       | 1.776s      |
+| 3000      | 100000    | 1          | 1          | 9.46s       |
+| 3000      | 100000    | 10         | 10         | 10.533s     |
+| 3000      | 100000    | 100        | 100        | 11.029s     |
+| 3000      | 100000    | 1000       | 1000       | 11.984s     |
```

Comment:
@wesley-weiming 

In https://github.com/tiktok/project-impact-graph/pull/3/files I've added some instrumentation to count the number of times each part of the loop is executed.

Here's one of the test cases:
```js
[
  {
    nodeCount: 2000,
    edgeCount: 10000,
    pathCountA: 1000,
    pathCountB: 1000,
    hasImpactIntersection: true,
    executeTime: '1.513s'
  }
]
{
  _integrateExcludedGlobs: 1,
  _integrateExcludedGlobs2: 2000,
  _validatePaths: 2,
  _validatePaths2: 4008000,
  lookUpProjectNamesByPathList: 2,
  lookUpProjectNamesByPathList2: 4000000,
  lookUpProjectNamesByPathList3: 4000000,
  lookUpProjectNamesByPathList4: 5780,
  getProjectImpactByProjectNames: 2,
  getProjectImpactByProjectNames2: 4000,
  getProjectImpactByProjectNames3: 24000,
  hasImpactIntersection: 1
}
```
The `nodeCount` is the number of projects, and `pathCountA` and `pathCountB` are the "before" and "after" lists of paths from the diff.  The bottleneck of this algorithm seems to be 4,000,000 which is `O(numberOfPaths * numberOfProjects)`.

But notice that our project paths have a well-behaved structure, for example:

- `INCLUDE` `apps/my-app/**`
- `EXCLUDE`  `apps/my-app/README.md`
- `INCLUDE` `apps/my-app2/**`
- `EXCLUDE`  `apps/my-app2/README.md`
- `INCLUDE` `libraries/my-lib3/**`
- `EXCLUDE`  `libraries/my-lib3/README.md`
- `INCLUDE` `libraries/my-lib4/**`
- `EXCLUDE`  `libraries/my-lib4/README.md`
- `INCLUDE` `libraries/my-lib4/bad-nested-project/**`
- `EXCLUDE`  `libraries/my-lib4/bad-nested-project/README.md`

Even if we permit projects to be nested under other project folders, the prefixes of these globs still form a tree.  For an input path given `libraries/my-lib3/src/index.ts`, imagine an `O(n*log(n))` algorithm that would cheaply walk down `libraries` -> `my-lib3`, and then need to test only 2 globs `**` and `README.md`.

This idea is similar to [rush-lib/src/logic/LookupByPath.ts](https://github.com/microsoft/rushstack/blob/main/libraries/rush-lib/src/logic/LookupByPath.ts).